src/core/thread/mle.cpp

```diff
@@ -1434,24 +1436,40 @@ otError Mle::AppendAddressRegistration(Message &aMessage, AddressRegistrationMod
         VerifyOrExit(counter < OPENTHREAD_CONFIG_MLE_IP_ADDRS_TO_REGISTER, OT_NOOP);
     }
 
-    // For sleepy end device, register external multicast addresses to the parent for indirect transmission
-    if (!IsRxOnWhenIdle())
+    // Append external multicast address.
+    // For sleepy end device, register external multicast addresses to the parent for indirect transmission.
+    // Since Thread 1.2, non-sleepy MED would depend on its Thread 1.2 parent as the proxy to perform
+    // Multicast Listener Report, it registers external multicast addresses of scope larger than realm
+    // local as well.
+    while (Get<ThreadNetif>().GetNextExternalMulticast(iterator, address) == OT_ERROR_NONE)
     {
-        uint8_t      iterator = 0;
-        Ip6::Address address;
-
-        // append external multicast address
-        while (Get<ThreadNetif>().GetNextExternalMulticast(iterator, address) == OT_ERROR_NONE)
+        if (IsRxOnWhenIdle())
         {
-            entry.SetUncompressed();
-            entry.SetIp6Address(address);
-            SuccessOrExit(error = aMessage.Append(&entry, entry.GetLength()));
-            length += entry.GetLength();
-
-            counter++;
-            // only continue to append if there is available entry.
-            VerifyOrExit(counter < OPENTHREAD_CONFIG_MLE_IP_ADDRS_TO_REGISTER, OT_NOOP);
+#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+            // For Thread 1.2 MED, skip MA registration if the parent is Thread 1.1 device.
+            if (GetParent().IsThreadVersion1_1())
+            {
+                break;
+            }
+            else if (!address.IsMulticastLargerThanRealmLocal())
+            {
+                // For Thread 1.2 MED, skip multicast address with scope not larger than realm local when registering.
+                continue;
+            }
+#else
+            // For Thread 1.1 MED, skip any multicast address.
+            break;
+#endif
```

Comment:
Can we consider moving some the common checks outside of the `while` loop.

I think something like this may work:
```
    // Append external multicast addresses.  For sleepy end device,
    // register all external multicast addresses with the parent for
    // indirect transmission. Since Thread 1.2, non-sleepy MED should
    // also register external multicast addresses of scope larger than
    // realm with a 1.2 or higher parent.

    if (!IsRxOnWhenIdle()  
    #if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
        || !GetParent().IsThreadVersion1_1()
    #endif
        )
    {
        while (Get<ThreadNetif>().GetNextExternalMulticast(iterator, address) == OT_ERROR_NONE)
        {
    #if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
            // For Thread 1.2 MED, skip multicast address with scope not
            // larger than realm local when registering
            if (IsRxOnWhenIdle() && !address.IsMulticastLargerThanRealmLocal())
            {
                continue;
            }
    #endif  

        ...

        }
    }
```

Reply:
no strong preference. update as your suggestion.